### ISSUE TITLE
API similar to pytorch for loading data

### ISSUE THREAD

**Issue by dnaq** [keras-team/keras#7359]
I've been using keras for some projects where I've needed each batch to contain multiple samples of each class, e.g. each batch should contain twenty classes and each class should have four samples. I can't really use the standard `ImageDataGenerators` `flow_from_directory` for this since it returns either the images in order or the images in random order.

However, pytorch/torchvision has a really nice data loading api where you have the following abstractions.

1. A dataset that includes all images/labels for your data. The dataset doesn't need to store its data in memory, examples of datasets are `ImageFolder`, `MNIST`, and so forth.
2. A sampler that defines how to draw samples from the dataset, a sampler could visit all items in the dataset in order, or it could draw items randomly or in some user-defined way.
3. A sequence of transforms that define how to transform the dataset.
4. A dataloader that draws batches from a dataset by the order defined by the sampler and yields minibatches of the dataset. The dataloader is responsible for multi-threading, pinning memory and so forth.

So for my use case I just need to write my own sampler, and then I can use the rest of the API transparently instead of having to write my own multi-threaded `ImageDataGenerator`-workalike.

Having written my own dataloaders, and having used tensorflows queues I must say that I'm quite impressed with the simplicity of pytorch in this case, and being that Keras is a really nice high-level framework for deep learning I think a similar API would be of use here too.

So my code would be something like:

```python
class ClassSampler(Sampler):
    def __init__(self, dataset, num_samples):
        super(ClassSampler, self).__init__()
        self.dataset = dataset
        self.num_samples = num_samples
        self.class_imgs = defaultdict(list)
        self.current_class_imgs = defaultdict(list)
        for img_idx, (_, class_) in enumerate(self.dataset.imgs):
            self.class_imgs[class_].append(img_idx)

    def __iter__(self):
        classes = list(self.class_imgs.keys()):
        random.shuffle(classes)
        for c in classes:
            for _ in range(self.num_samples):
                if not self.current_class_imgs[c]:
                    self.current_class_imgs[c] = self.class_imgs[c][:]
                    random.shuffle(self.current_class_imgs[c])
                yield self.current_class_imgs[c].pop()

    def __len__(self): return len(self.dataset.imgs)

trainset = ImageFolder(path)
sampler = ClassSampler(trainset, 4)
train_transform = transforms.Compose([transforms.RandomCrop((160, 160)), transforms.RandomHorizontalFlip(), transforms.Lambda(preprocess_input)])
trainloader = DataLoader(trainset, batch_size=128, num_workers=8, drop_last=True, sampler=sampler)

model.fit_generator(trainloader, ...)
```

Does that make sense? There is a lot more to read on this issue in the pytorch documentation, for example [this tutorial](http://pytorch.org/tutorials/beginner/data_loading_tutorial.html).

**Comment by Adityakunwar** [comment]
@dnaq have you tried this API on a custom dataset? I am looking for something like Pytorch's Dataloader.

**Comment by hieuttcse** [comment]
@Adityakunwar, You may want to try [Sequence](https://keras.io/utils/#sequence)
